sparse-traits/examples/polynomials.rs

```diff
@@ -0,0 +1,138 @@
+        Polynomial {
+            monomial_coeffs: monomial_coeffs.to_owned(),
+        }
+    }
+    fn eval(&self, x: f64) -> f64 {
+        self.monomial_coeffs.iter().rev().fold(0., |r, c| r * x + c)
+    }
+}
+
+#[derive(Debug)]
+struct PolynomialView<'a> {
+    monomial_coeffs: &'a [f64],
+}
+struct PolynomialViewMut<'a> {
+    monomial_coeffs: &'a mut [f64],
+}
+
+impl Element for Polynomial {
+    type Space = PolynomialSpace;
+    type View<'a> = PolynomialView<'a> where Self: 'a ;
+    type ViewMut<'a> = PolynomialViewMut<'a> where Self: 'a;
+
+    fn view<'a>(&'a self) -> Self::View<'a> {
+        PolynomialView {
+            monomial_coeffs: &self.monomial_coeffs,
+        }
+    }
+    fn view_mut<'a>(&'a mut self) -> PolynomialViewMut<'a> {
+        PolynomialViewMut {
+            monomial_coeffs: &mut self.monomial_coeffs,
+        }
+    }
+}
+
+struct PointwiseEvaluatorSpace;
+impl LinearSpace for PointwiseEvaluatorSpace {
+    type F = f64;
+    type E = PointwiseEvaluate;
+}
+impl DualSpace for PointwiseEvaluatorSpace {
+    type Space = PolynomialSpace;
+
+    fn dual_pairing(&self, x: &Self::E, p: &<Self::Space as LinearSpace>::E) -> Result<Self::F> {
+        Ok(x.scale * p.eval(x.x))
+    }
+}
+
+struct PointwiseEvaluate {
+    x: f64,
+    scale: f64,
+}
+impl PointwiseEvaluate {
+    fn new(x: f64) -> Self {
+        PointwiseEvaluate { x, scale: 1. }
+    }
+}
+
+impl Element for PointwiseEvaluate {
+    type Space = PointwiseEvaluatorSpace;
+    type View<'a> = &'a PointwiseEvaluate where Self: 'a;
+    type ViewMut<'a> = &'a mut PointwiseEvaluate where Self: 'a;
+    fn view<'a>(&'a self) -> Self::View<'a> {
+        &self
+    }
+    fn view_mut<'a>(&'a mut self) -> Self::ViewMut<'a> {
+        self
+    }
+}
+
+#[derive(Debug)]
+struct Derivative;
+impl OperatorBase for Derivative {
+    type Domain = PolynomialSpace;
+    type Range = PolynomialSpace;
+    fn as_apply(&self) -> Option<&dyn AsApply<Domain = Self::Domain, Range = Self::Range>> {
+        Some(self)
+    }
+}
+impl AsApply for Derivative {
+    fn apply(&self, p: PolynomialView, dp: PolynomialViewMut) -> Result<()> {
+        for (i, c) in p.monomial_coeffs[1..].iter().enumerate() {
+            dp.monomial_coeffs[i] = (1. + i as f64) * c;
+        }
+        dp.monomial_coeffs[p.monomial_coeffs.len() - 1] = 0.;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        Derivative, PointwiseEvaluate, PointwiseEvaluatorSpace, Polynomial, PolynomialSpace,
+    };
+    use sparse_traits::{AsApply, DualSpace, Element, OperatorBase, Result};
+
+    #[test]
+    fn test_poly_eval() {
+        let p = Polynomial::from_monomial(&[1., 2., 3.]);
+        assert_eq!(p.eval(2.), 17.);
+    }
+
+    #[test]
+    fn test_dual() -> Result<()> {
+        let ds = PointwiseEvaluatorSpace;
+        let p = Polynomial::from_monomial(&[1., 2., 3.]);
+        let n = PointwiseEvaluate::new(2.);
+        let r = ds.dual_pairing(&n, &p)?;
```

Comment:
I think needing to refer to the dual space explicitly is undesirable. I'd rather see `let r = n.dual_pairing(&p)`.

Reply:
Should it be possible to use views here? I'd think yes.

Reply:
> I think needing to refer to the dual space explicitly is undesirable. I'd rather see `let r = n.dual_pairing(&p)`.

This is easy to do since Elements refer the underlying space.

Reply:
> Should it be possible to use views here? I'd think yes.

Yes. It was an oversight by me.

Reply:
Would you make a `DualElement`? Or would all `Element`s have a `dual_pairing` method? If so, what is the dual of `PolynomialSpace`?

BTW, this is the Riesz representation of $\delta(x - 0.4) \in P_{50}^*$ in the dual space of $P_{50}$, the polynomials of max degree less than 50. It takes work to construct and can't be constructed when the max degree is not bounded. Ignoring computation, there's the issue that we sometimes work with spaces that are not complete (like polynomials). I don't know of a pragmatic way to preserve the symmetry that exists in analysis.
![image](https://user-images.githubusercontent.com/3303/203903218-6b95cd17-15c8-4e52-b983-aaa8f8fe5ab2.png)